packages/compiler/src/server/tspconfig/completion.ts

```diff
@@ -211,3 +354,240 @@ export async function provideTspconfigCompletionItems(
+ * @param source source text
+ * @param nodePosRange current node position range, [startPos, endPos, nodeEndPos]
+ * @param curPos current cursor position
+ * @param tspConfigPosition original position, see {@link Position}
+ * @returns TextEdit object or undefined
+ */
+function getFullCompletionItemValAndEditPosition(
+  newText: string,
+  source: string,
+  nodePosRange: number[],
+  curPos: number,
+  tspConfigPosition: Position,
+): TextEdit | undefined {
+  const [startPos, endPos] = nodePosRange;
+
+  if (curPos >= startPos && curPos <= endPos) {
+    return TextEdit.replace(
+      Range.create(
+        Position.create(tspConfigPosition.line, tspConfigPosition.character - source.length),
+        Position.create(tspConfigPosition.line, tspConfigPosition.character + (endPos - curPos)),
+      ),
+      newText,
+    );
+  }
+  return undefined;
+}
+
+/**
+ * Get the common CompletionItem object, which value is included in quotes
+ * @param labelName The value of the label attribute of the CompletionItem object
+ * @param sourceType Input quote type
+ * @param source Entered text
+ * @param description The value of the documentation attribute of the CompletionItem object
+ * @param tspConfigPosition Input location object, see {@link Position}
+ * @param nodePosRange current node position range, [startPos, endPos, nodeEndPos]
+ * @param curPos current cursor position
+ * @returns CompletionItem object or undefined
+ */
+function createContainingQuatedValCompetionItem(
+  labelName: string,
+  sourceType: string,
+  source: string,
+  description: string,
+  tspConfigPosition: Position,
+  nodePosRange: number[],
+  curPos: number,
+): CompletionItem | undefined {
+  // Generate new text
+  const newText = getCompletionItemInsertedValue(sourceType, labelName);
+  const textEdit = getFullCompletionItemValAndEditPosition(
+    newText,
+    source,
+    nodePosRange,
+    curPos,
+    tspConfigPosition,
+  );
+  if (textEdit !== undefined) {
+    return {
+      label: labelName,
+      kind: CompletionItemKind.Field,
+      documentation: description,
+      textEdit: textEdit,
+    };
+  }
+  return undefined;
+}
+
+/**
+ * Find a set of dirs/files with the same suffix
+ * @param rootPath The absolute input path or relative path of the current configuration file
+ * @param fileExtension  File extension, such as ".yaml" or ".tsp"
+ * @param configFile absolute path of the current configuration file
+ * @returns dir/file array
+ */
+async function findFilesOrDirsWithSameExtension(
+  rootPath: string,
+  fileExtension: string,
+  configFile: string = "",
+): Promise<string[]> {
+  const exclude = ["node_modules", "tsp-output", ".vs", ".vscode"];
+
+  const files: string[] = [];
+  try {
+    // When reading the content under the path, an error may be reported if the path is incorrect.
+    // Exclude the current configuration file, compare in absolute paths
+    (await fs.readdir(rootPath, { withFileTypes: true }))
```

Comment:
please use host.compilerhost to do these IO operations. usually we won't call IO operatoin directly. same comment for other IO operation.